### PR TITLE
Fixed generic distinct conversion proc calls

### PR DESCRIPTION
**Pull request by beef331**

### CODE REVIEW

File: compiler/jsgen.nim [hunk 1]
```diff
@@ -1288,75 +1288,89 @@ template isIndirect(x: PSym): bool =
...
-proc genAddr(p: PProc, n: PNode, r: var TCompRes) =
-  case n[0].kind
-  of nkSym:
-    let s = n[0].sym
-    if s.loc.r == nil: internalError(p.config, n.info, "genAddr: 3")
-    case s.kind
-    of skParam:
-      r.res = s.loc.r
-      r.address = nil
+proc genSymAddr(p: PProc, n: PNode, typ: PType, r: var TCompRes) =
```
Comment (saem) [inline compiler/jsgen.nim:1291]: Can haz doc comments for the uninitiated?

File: compiler/jsgen.nim [hunk 2]
```diff
@@ -1288,75 +1288,89 @@ template isIndirect(x: PSym): bool =
...
+    r.typ = etyNone
+  of skVar, skLet, skResult:
+    r.kind = resExpr
+    let jsType = mapType(p):
+      if typ.isNil:
+        n.typ
+      else:
+        typ
+    if jsType == etyObject:
+      # make addr() a no-op:
```
Comment (saem) [inline compiler/jsgen.nim:1310]: It's that because they're already refs in JS land?

File: compiler/jsgen.nim [hunk 3]
```diff
@@ -1288,75 +1288,89 @@ template isIndirect(x: PSym): bool =
...
-      # namely toOpenArray(a, 1, 3)
+      gen(p, n, r)
+      #internalError(p.config, n.info, "genAddr: 4 " & renderTree(n))
+  else: internalError(p.config, n.info, $("genAddr: 2", s.kind))
+
+proc genAddr(p: PProc, n: PNode, r: var TCompRes) =
+  if n.kind == nkSym:
+    genSymAddr(p, n, nil, r)
+  else:
+    case n[0].kind
```
Comment (saem) [inline compiler/jsgen.nim:1333]: Why do we look at the first child?

File: compiler/types.nim [hunk 4]
```diff
@@ -1121,15 +1121,15 @@ proc sameTypeAux(x, y: PType, c: var TSameTypeClosure): bool =
     case c.cmp
     of dcEq: return false
     of dcEqIgnoreDistinct:
-      while a.kind == tyDistinct: a = a[0]
-      while b.kind == tyDistinct: b = b[0]
+      a = a.skipTypes({tyDistinct, tyGenericInst})
```
Comment (saem) [inline compiler/types.nim]: Skip types is much better!

Just wondering if we're missing any others? 🤔

File: tests/converter/texplicit_conversion.nim [hunk 5]
```diff
@@ -11,3 +11,9 @@ converter toInt(s: string): int =
 
 let x = (int)"234"
 echo x
+
+block: # Test for nkconv
```
Comment (saem) [inline tests/converter/texplicit_conversion.nim]: A few words describing what the spec should be/is would help teach people the spec and/or build it out as we go in an incremental fashion.

File: tests/distinct/tdistinct.nim [hunk 6]
```diff
@@ -140,6 +141,22 @@ block tRequiresInit:
     let s = "test"
     doAssert s == "test"
 
+block: #17322
```
Comment (saem) [inline tests/distinct/tdistinct.nim]: I'd prefer we keep nim issue references only for deep dive, but being in information into nimskull.

**Review by saem — CHANGES_REQUESTED** [review]

File: compiler/types.nim (hunk 4 above)
Comment (beef331) [reply compiler/types.nim]: Added a `skipDistincts` to make it easier to handle all cases when wanting to skip distincts, still need to replace any uses of skipTypes with it.

**Review by saem — APPROVED** [review]

**Pull request merged**